gh-101561: Add typing.override decorator

stroxler opened the pull request:
The code is pulled almost unchanged from `typing_extensions`:

- [decorator implementation](https://github.com/python/typing_extensions/blob/main/src/test_typing_extensions.py#L168-L202)
- [tests](https://github.com/python/typing_extensions/blob/main/src/test_typing_extensions.py#L168-L202)

Question that came up was where to put the type variable and what to name it:
- In `typing_extensions` the type variables use underscore prefixes to indicate that they are not public, but `typing` doesn't appear to use that convention so I named the type variable `F`.
- I couldn't put the type variable next to all of the other ones (used in generic container definitions) because it relies on `Callable` which is defined below. If this is a problem I could correct it using a forward reference, or we could try to rearrange the module to put special forms above type variables.

## Testing the code:

First set up the repo, following instructions
at https://devguide.python.org/ by running:
```
./configure --with-pydebug && make -j
```

Then run the typing tests:
```
./python -m test test_typing -v
```

I ran the full test suite with
```
./python -m test -j3
```
and it came back clean except for a `test_grp` failure which I seem to get on trunk as well - likely something in my build is misconfigured but I'm pretty sure it is unrelated to the changes here.

<!-- gh-issue-number: gh-101561 -->
* Issue: gh-101561
<!-- /gh-issue-number -->


## Make sure the docs build:

```
cd Doc
make venv
make html
```

stroxler committed:
PEP 698: Add typing.override decorator

Testing:

First set up the repo, following instructions
at https://devguide.python.org/ by running:
```
./configure --with-pydebug && make -j
```

Then run the typing tests:
```
./python -m test test_typing -v
```

I ran the full test suite with
```
./python -m test -j3
```
and it came back clean except for a `test_grp` failure which I seem to
get on trunk as well - likely something in my build is misconfigured
but I'm pretty sure it is unrelated to the changes here.

stroxler committed:
Add missing `For example::` in the rst doc

AlexWaygood reviewed (commented):
Thanks! Haven't looked in depth yet at all, but we'll also need:
- A news entry, as @bedevere-bot says
- An entry in `Doc/whatsnew/3.12.rst`
- (optionally) you can add your name to `Misc/ACKS`

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

JelleZijlstra commented on Doc/library/typing.rst:
```diff
@@ -2722,6 +2724,42 @@ Functions and decorators
+   The decorator will set the ``__override__`` attribute to ``True`` on
+   the decorated object. Thus, a check like
+   ``if getattr(obj, "__final__", False)`` can be used at runtime to determine
+   whether an object ``obj`` has been marked as final.  If the decorated object
```
```suggestion
   ``if getattr(obj, "__override__", False)`` can be used at runtime to determine
   whether an object ``obj`` has been marked as an override.  If the decorated object
```

stroxler replied:
Yikes, you can tell I copied this to get the formatting right.

stroxler committed:
Add whatsnew entry, as per Alex Waygood's suggestion

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

stroxler committed:
Update Doc/library/typing.rst

Correct `s/final/override` (from reusing a similar block of documentation on the `@final` decorator)

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

stroxler committed:
Add ACKS entry, fix rst errors

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

stroxler committed:
Fix title underline length in whatisnew doc.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

AlexWaygood commented on Misc/NEWS.d/next/Library/2023-02-04-16-35-46.gh-issue-101561.Xo6pIZ.rst:
```diff
@@ -0,0 +1,5 @@
+
+This decorator indicates to static type checkers that they should verify the method overrides an attribute or method of the same name in a base class. This is useful for catching bugs when base class methods are renamed but some children are not updated accordingly.
+
+See :pep:`698` for details.
```
```suggestion
See :pep:`698` for details. Patch by Steven Troxler.
```

stroxler committed:
Update Misc/NEWS.d/next/Library/2023-02-04-16-35-46.gh-issue-101561.Xo6pIZ.rst

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

TeamSpen210 commented on Lib/typing.py:
```diff
@@ -3448,3 +3452,40 @@ def decorator(cls_or_fn):
+
+
+
+def override(__arg: F) -> F:
```
`typing_extensions` has to use this workaround since it supports Python 3.7, but `typing` can just use a positional-only arg. Might also be better to name it something more meaningful like `func` or `method`?

stroxler committed:
Use a positional-only argument named `method`

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

JelleZijlstra commented on Doc/whatsnew/3.12.rst:
```diff
@@ -350,6 +350,14 @@ tempfile
+* Add :func:`typing.override`, an override decorator telling to static type
+  checkers to verify that a method overrides some method or attribute of the
+  same name on a base class, as per :pep:`698` (Contributed by Steven Troxler in
+  :gh:`101564`).
```
```suggestion
  same name on a base class, as per :pep:`698`. (Contributed by Steven Troxler in
  :gh:`101564`.)
```
punctuation nit

JelleZijlstra commented on Misc/NEWS.d/next/Library/2023-02-04-16-35-46.gh-issue-101561.Xo6pIZ.rst:
```diff
@@ -0,0 +1,5 @@
+
+This decorator indicates to static type checkers that they should verify the method overrides an attribute or method of the same name in a base class. This is useful for catching bugs when base class methods are renamed but some children are not updated accordingly.
+
+See :pep:`698` for details. Patch by Steven Troxler.
```
```suggestion
Add a new decorator :func:`typing.override`. See :pep:`698` for details. Patch by Steven Troxler.
```
This can be kept more concise. The What's New can have more of an explanation.

JelleZijlstra reviewed (approved):
Sorry for dropping this; I have a few nits but I'm happy to merge the PR. I'll leave it open for a few days in case Guido does want to take a look himself.

stroxler committed:
Apply suggestions from code review

Thanks, looks good

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>
